File: app/api/csrf.py
from flask import request


def csrf_token_from_request():
    """
    The CSRF token the client sent with this request.

    It has to arrive in the X-CSRFToken header rather than in the cookie. The
    browser attaches the cookie to cross-site requests by itself, so reading the
    token from it meant the check could never fail; the same origin policy stops
    another site reading our cookie to set the header, so this half of the
    double submit is the half that actually proves the request came from us.

    Returns '' rather than raising when the header is absent, so a request
    without one fails form validation with clean JSON instead of blowing up with
    a BadRequestKeyError.
    """
    return request.headers.get('X-CSRFToken', '')

File: react-app/src/csrf.js
// The server hands us a CSRF token in a readable cookie and expects it back in
// the X-CSRFToken header on anything that changes state. The browser attaches
// the cookie to cross-site requests on its own, so the cookie alone proves
// nothing; another origin cannot read it to set the header, so the header does.
export const csrfToken = () => {
	const match = document.cookie.match(/(?:^|;\s*)csrf_token=([^;]*)/);
	return match ? decodeURIComponent(match[1]) : "";
};

// Headers for a request that changes something on the server. Pass the headers
// the request already needs and this adds the token to them.
export const csrfHeaders = (headers = {}) => ({
	...headers,
	"X-CSRFToken": csrfToken(),
});
